src/frame_buffer/frame_buffer.rs

```diff
@@ -13,80 +17,59 @@ pub type DropMetadata = &'static (dyn Any + Send + Sync);
 /// Buffers are released in their destructor methods provided by the buffer allocator.
 ///
 /// [`FrameBuffer`] is a smart pointer dereferencing a bytes slice.
-pub struct FrameBuffer {
-    frame: &'static mut [u8],
-    drop_fn: Option<DropFunction>,
-    drop_md: DropMetadata,
+pub struct FrameBuffer<'a> {
+    allocator: FrameAllocators<'a>,
+    frame: &'a mut [u8],
 }
 
-impl FrameBuffer {
-    /// Creates a new buffer
-    ///
-    /// ---
-    /// **NOTE** this function is to be used only by [frame buffer
-    /// allocators](super::frame_allocator::FrameAllocator) or in tests.
-    ///
-    /// ---
-    ///
-    /// The buffer structure points to the `frame` memory slice provided by the caller. This slice
-    /// is considered owned by the [`FrameBuffer`] until the `FrameBuffer` is dropped. While the
-    /// slice is owned by the `FrameBuffer` it shall not be accessed in any other way than by
-    /// dereferencing the `FrameBuffer`.
-    ///
-    /// Releasing the memory slice for other use is indicated by the `FrameBuffer` by calling the
-    /// `drop_fn` function passing a reference to the buffer being freed and a caller-defined
-    /// `drop_md` reference.
-    pub fn new(
-        frame: &'static mut [u8],
-        drop_fn: Option<DropFunction>,
-        drop_md: DropMetadata,
-    ) -> Self {
-        Self {
-            frame,
-            drop_fn,
-            drop_md,
-        }
+impl<'a> FrameBuffer<'a> {
+    pub fn new(allocator: FrameAllocators<'a>, frame: &'a mut [u8]) -> Self {
+        Self { allocator, frame }
     }
 }
 
-impl Debug for FrameBuffer {
+impl<'a> Debug for FrameBuffer<'a> {
     fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), core::fmt::Error> {
         f.debug_struct("FrameBuffer")
             .field("frame", &self.frame)
             .finish()
     }
 }
 
-impl Deref for FrameBuffer {
+impl<'a> Deref for FrameBuffer<'a> {
     type Target = [u8];
 
     fn deref(&self) -> &Self::Target {
         self.frame
     }
 }
 
-impl DerefMut for FrameBuffer {
+impl<'a> DerefMut for FrameBuffer<'a> {
     fn deref_mut(&mut self) -> &mut Self::Target {
         self.frame
     }
 }
 
-impl Drop for FrameBuffer {
+impl<'a> Drop for FrameBuffer<'a> {
     fn drop(&mut self) {
-        if let Some(drop_fn) = self.drop_fn {
-            drop_fn(self, self.drop_md)
+        match self.allocator {
+            #[cfg(test)]
+            FrameAllocators::Dummy => (),
+            FrameAllocators::SingleFrameAllocator(allocator) => {
+                SingleFrameAllocator::release_frame(allocator, self)
```

Comment:
This requires FrameBuffer to be aware of correct usage of allocators. It should be the other way around. It's the allocator who should instruct FrameBuffer what to do to release a frame. Such dependencies end in shotgun surgery anti-pattern I hope to avoid in this pet project :)

Reply:
If you have correct `release_frame` for given allocator then there is no way to get it wrong, especially that allocator type is unique for each variant. Same goes for any needed metadata that should be bundled in variant as well. If you have `SingleJumboFrameAllocator`, there is no way to call it, having `SingleFrameAllocator` variant.

Reply:
Taking this into account, I can't see a way to add new allocator and get it wrong - code won't compile unless you have done everything. You can try it yourself by adding `SingleJumboFrameAllocator` with bigger frame size than this one on this branch.

Reply:
I don't mean overseeing mistakes in the code. I'm sure compiler will catch that. My concern is this change increases coupling between allocator and frames, and decreases cohesion of the frame_buffer module.